civicrm_admin_ui: convert Browse mailing

samuelsov opened the pull request:
Overview
----------------------------------------
Convert Mailing browse page to Search Kit + FormBuilder


Before
----------------------------------------
![ksnip_20230614-222956](https://github.com/civicrm/civicrm-core/assets/372004/a791b486-2939-4199-a286-85138832353a)

After
----------------------------------------
![ksnip_20230614-201543](https://github.com/civicrm/civicrm-core/assets/372004/3421ed27-5e76-404d-aaaa-b167645bbb49)

Comments
----------------------------------------

Work in progress:
- fix "Created or Sent by"
- make all the route works:
   - action will stop working if we use civicrm/mailing/browse
   - do some extra route to replace default civicrm/mailing/browse/scheduled, civicrm/mailing/browse/unscheduled?
- add language column/filter only if multilingual?
- confirm that the colors for scheduled (green) and paused (red) mailing to make them more obvious are desirable?

samuelsov committed:
AdminUI: convert Browse mailing

differenciate draft/scheduled/archived

larssandergreen commented on ext/civicrm_admin_ui/ang/afsearchMailingBrowseScheduled.aff.html:
```diff
@@ -0,0 +1,11 @@
+<div af-fieldset="">
+  <div class="af-container af-layout-inline">
+    <af-field name="name" />
+    <af-field name="created_date" defn="{input_type: 'Select', search_range: true, options: [{id: '{}', label: 'Choose Date Range'}, {id: 'this.day', label: 'Today'}, {id: 'previous.day', label: 'Yesterday'}, {id: 'ending.week', label: 'Last 7 days including today'}, {id: 'ending.month', label: 'Last 30 days including today'}, {id: 'ending_2.month', label: 'Last 60 days including today'}, {id: 'e ...
```
There must be a way to load the options directly here.

larssandergreen commented on ext/civicrm_admin_ui/ang/afsearchMailingBrowseScheduled.aff.html:
```diff
@@ -0,0 +1,11 @@
+  <div class="af-container af-layout-inline">
+    <af-field name="name" />
+    <af-field name="created_date" defn="{input_type: 'Select', search_range: true, options: [{id: '{}', label: 'Choose Date Range'}, {id: 'this.day', label: 'Today'}, {id: 'previous.day', label: 'Yesterday'}, {id: 'ending.week', label: 'Last 7 days including today'}, {id: 'ending.month', label: 'Last 30 days including today'}, {id: 'ending_2.month', label: 'Last 60 days including today'}, {id: 'e ...
+    <af-field name="from_name" defn="{label: 'Created or Send by', input_attrs: {}}" />
```
Could we do a contact entity ref select here?

larssandergreen replied:
Also 'Created or Sen**t** by'

larssandergreen commented on ext/civicrm_admin_ui/ang/afsearchMailingBrowseScheduled.aff.json:
```diff
@@ -0,0 +1,19 @@
+    "title": "Find Mailings",
+    "icon": "fa-list-alt",
+    "server_route": "civicrm/mailing/browse/scheduled",
+    "permission": "access CiviCRM",
```
access CiviMail?

larssandergreen commented on ext/civicrm_admin_ui/managed/SavedSearch_Mailing_Browse_Scheduled.mgd.php:
```diff
@@ -0,0 +1,327 @@
+              'type' => 'field',
+              'key' => 'scheduled_id.display_name',
+              'dataType' => 'String',
+              'label' => E::ts('Send By'),
```
'Sen**t** by'

larssandergreen commented:
Looks good!

Like the colours. I think yellow would make sense for paused rather than red (I might assume that red means it has been cancelled or in some kind of error state), but not a big deal.

Agreed on hiding language if not multilingual. Also need to show Campaign if enabled (search and column).

samuelsov commented:
Done:
- [x] access CiviMail
- [x] Sent by
- [x] Language and campaign column hidden by default
- [x] yellow line for the paused mailing
- [x] added simplified view for Draft
- [x] removed the `Is archived` filter by having 2 different pages (only one SearchKit but 2 formbuilders)

Remaining tasks:
- [ ]  we need to create alternate actions for disabling, resume, canceling that don't quit the current page, only refresh it
- [ ]  I see that sms use the same search, so we need to ensure it is still working
- [ ]  make `ng-if` work for optional language and campaign filters

colemanw commented on ext/civicrm_admin_ui/managed/SavedSearch_Mailing_Browse_Scheduled.mgd.php:
```diff
@@ -0,0 +1,352 @@
+    'dataType' => 'Timestamp',
+    'label' => E::ts('Completed'),
+    'sortable' => TRUE,
+  ]
```
`civilint` is complaining about this.
```suggestion
  ],
```

colemanw commented:
> @samuelsov removed the `Is archived` filter by having 2 different pages (only one SearchKit but 2 formbuilders)

What was the reason for this change? From a developer POV it seems simpler to have just one form. Is there a difference from the user POV? (Note that if the reason was to have 2 nav menu items, we **can** set filters from the url so it could be possible to create 2 links to the same form with 2 different filter values set).

samuelsov commented:
@colemanw no reason that I can remember except trying to replace the 2 existing routes in case it is used elsewhere in the code.

mlutfy commented:
I committed the syntax fixes reported by Coleman and "Sent By"  reported by Lars.

About the "is archived", I think Samuel meant that we have 3 lists: Drafts, Scheduled/Sent, Archived. It can be the same form, with arguments. The "is archived" option is not really relevant if we keep those 3 menu items/lists.

Then there is the question of whether we replace routes or provide new routes. I've seen sometimes it either one or the other. It's easier to add new routes, but then there may be other areas in the UI that link to specific URLs, so admins would be confused to suddenly be on the old UI?

In other words, in this case, we would be replacing 3 legacy routes by a single route, with URL arguments? (or maybe just 2 routes: Sent vs Archived)

colemanw commented:
Thanks @mlutfy - I think I can work up a little `hook_civicrm_alterMenu()` magic so that all existing links will point to the new single route. (so "yes" to your question - we can replace all 3 with a single route with arguments).

colemanw committed:
AdminUI - Update mailing screen

colemanw committed:
AdminUI - Alter menus and routes for mailing screen

colemanw commented:
I'm working on this and have broken out two smaller PRs:
- #28019 
- #28019

samuelsov commented:
@colemanw A round of testing:

1. Pause/Resume/Cancel actions are not working
2. There is no menu action to Archive a mailing
3. The `is_draft` filter from url doesn't seem to work (there seems to be a value but neither `Yes` nor `No`):
![ksnip_20231108-102017](https://github.com/civicrm/civicrm-core/assets/372004/15016137-9fcd-4a29-a059-556ef524cade)
4. After adding a mailing, we go to a generic pages with no selected filter. Ideally, we should keep the current behavior, i.e. :
    - always select Archived = `No` by default
    - if we save a draft, select `Draft` filter to `Yes`
    - if we schedule the mailing, select `Draft` filter to `No`
5. This breaks the `Find Mass Sms` screen as it uses the same url with a parameter `sms=1`.

colemanw commented:
@samuelsov thanks for testing. Yes drafts are re-broken because we have reverted #28019 in favor of #27559 but that is still waiting for a piece from @totten.
So this PR is very close to finished but I'm going to change it back to draft status for now while we wait for those few missing pieces.

mlutfy commented:
Any thoughts on how to implement this item?

> "we need to create alternate actions for disabling, resume, canceling that don't quit the current page, only refresh it"

magnolia61 commented:
Slightly related: would it be an idea to call this page Mailing Dashboard? For consistency with the other menu entities?

shaneonabike commented:
Both of the issues above have been merged now I think. Can we do another round of testing and merge it? @samuelsov @colemanw

colemanw committed:
AdminUI - Use new 'status' column for mailing screen

mlutfy commented:
Excellent :tada:

I did a bit of poking around, and works for me.

I would add merge-on-pass, but we have to wait for #33561 to be merged first.

colemanw commented:
That only took 2 years! Full speed ahead! 🚀

colemanw commented:
@composerjk that was fixed with https://github.com/civicrm/civicrm-core/pull/34086
